### PR TITLE
fix(app): validator set update

### PR DESCRIPTION
**Pull request by aleem1314**
## Description

Closes: #XXXX

---

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

- [ ] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] added `!` to the type prefix if API or client breaking change
- [ ] targeted the correct branch (see [PR Targeting](https://github.com/cosmos/cosmos-sdk/blob/main/CONTRIBUTING.md#pr-targeting))
- [ ] provided a link to the relevant issue or specification
- [ ] followed the guidelines for [building modules](https://github.com/cosmos/cosmos-sdk/blob/main/docs/building-modules)
- [ ] included the necessary unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/main/CONTRIBUTING.md#testing)
- [ ] added a changelog entry to `CHANGELOG.md`
- [ ] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [ ] updated the relevant documentation or specification
- [ ] reviewed "Files changed" and left comments if necessary
- [ ] confirmed all CI checks have passed

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed `!` in the type prefix if API or client breaking change
- [ ] confirmed all author checklist items have been addressed 
- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

### CODE REVIEW

File: app/app.go [hunk 1]
```diff
@@ -591,9 +591,21 @@ func MakeCodecs() (codec.Codec, *codec.LegacyAmino) {
 // Name returns the name of the App
 func (app *RegenApp) Name() string { return app.BaseApp.Name() }
 
+const PATCH_HEIGHT = 600 // TODO: update patch height
+
+var voteInfos []abci.VoteInfo
```
Comment (aaronc) [inline app/app.go:596]: we can make this a field on `RegenApp`

File: app/app.go [hunk 2]
```diff
@@ -591,9 +591,21 @@ func MakeCodecs() (codec.Codec, *codec.LegacyAmino) {
...
+
 // BeginBlocker application updates every begin block
 func (app *RegenApp) BeginBlocker(ctx sdk.Context, req abci.RequestBeginBlock) abci.ResponseBeginBlock {
 	resp := app.mm.BeginBlock(ctx, req)
+
+	// initialize global tendermint validator set before patch height
+	if ctx.BlockHeight() < PATCH_HEIGHT {
```
Comment (aaronc) [inline app/app.go:603]:
```suggestion
	if ctx.BlockHeight() <= PATCH_HEIGHT {
```

**Review by aaronc — COMMENTED** [review]
I think the logic is not quite right here, what I think we need is:
* for all validators in app state (`app.StakingKeeper.GetAllValidators`), create a `ValidatorUpdate` with the correct power
* for all remaining validators in `[]VoteInfo` that are _not_ in app state, set their voting power to zero

```
var updates []abci.ValidatorUpdate
validators := app.StakingKeeper.GetAllValidators(ctx)
inValidatorSet := map[sdk.ConAddress]bool{}

for _, validator := range validators {
  updates = append(updates, abci.ValidatorUpdate{...})
  inValidatorSet[consAddr] = true
}

for _, voteInfo := range voteInfos {
  if !inValidatorSet[vote.Validator.Address] {
   // the big question is how to get the pubkey for a validator is no longer in state
    updates = append(abci.ValidatorUpdate{PubKey: pk, Power: 0})
  }
} 
```

File: app/app.go [hunk 3]
```diff
@@ -602,10 +616,101 @@ func (app *RegenApp) BeginBlocker(ctx sdk.Context, req abci.RequestBeginBlock) a
...
+				panic(err)
+			}
+
+			valCons, err := validator.GetConsAddr()
+			if err != nil {
+				panic(err)
+			}
+
+			if validator.Jailed {
+				for _, voteInfo := range app.votesInfo {
```
Comment (aleem1314) [inline app/app.go:648]: @aaronc returning validator which does not exist in tendermint state is halting the network with this error
```
Sep 15 09:35:24 regen-devnet-val-01 cosmovisor[19623]: 9:35AM ERR failed to apply block err="commit failed for application: error changing validator set: failed to find validator 001F90AE99B52F548845DA4D93755C3D4B829097 to remove" height=600 module=consensus
```

File: app/app.go [hunk 4]
```diff
@@ -602,10 +615,111 @@ func (app *RegenApp) BeginBlocker(ctx sdk.Context, req abci.RequestBeginBlock) a
...
+
+						// We need to reset the counter & array so that the validator won't be immediately slashed for downtime after unjail.
+						signInfo, found := app.SlashingKeeper.GetValidatorSigningInfo(ctx, valCons)
+						if found {
+							signInfo.MissedBlocksCounter = 0
+							signInfo.IndexOffset = 0
+							app.SlashingKeeper.SetValidatorSigningInfo(ctx, valCons, signInfo)
+							missedBlocks := app.SlashingKeeper.GetValidatorMissedBlocks(ctx, valCons)
+							for _, mb := range missedBlocks {
+								app.SlashingKeeper.SetValidatorMissedBlockBitArray(ctx, valCons, mb.Index, false)
```
Comment (aleem1314) [inline app/app.go:665]: This method `app.SlashingKeeper.clearValidatorMissedBlockBitArray` is private, so here I am resetting `ValidatorMissedBlockBitArray`.

**Comment by clevinson** [comment]
Closing this PR in favor of #1501 which was merged earlier today.

**Pull request closed without merging**